pkg/download/download.go

```diff
@@ -0,0 +1,186 @@
+package download
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/firmwareci/contest-client/pkg/env"
+	"github.com/jlaffaye/ftp"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// DownloadHttp downloads the file at the end of the Url via http/ https,
+// into the given file. Both pointers owned by caller.
+func DownloadHTTP(file *os.File, URL *url.URL) error {
+	resp, err := http.Get(URL.String())
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if _, err := io.Copy(file, resp.Body); err != nil {
+		return err
+	}
+	return nil
+}
+
+// DownloadFtp downloads the file at the end of the Url via ftp,
+// into the given file. Both pointers owned by caller.
+func DownloadFTP(file *os.File, URL *url.URL) error {
+	if URL.User == nil {
+		URL.User = url.UserPassword("anonymous", "anonymous")
+	}
+
+	connection, err := ftp.DialTimeout(URL.Hostname(), 60*time.Second)
+	if err != nil {
+		return err
+	}
+
+	defer connection.Quit()
+
+	pw, set := URL.User.Password()
+
+	if !set {
+		pw = "anonymous"
+	}
+
+	if err := connection.Login(URL.User.Username(), pw); err != nil {
+		return err
+	}
+
+	resp, err := connection.Retr(URL.Path)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(file, resp); err != nil {
+		return err
+	}
+	resp.Close()
+
+	return nil
+}
+
+// DownloadSFTP downloads the file at the end of the Url via sftp,
+// into the given file. Both pointers owned by caller.
+func DownloadSFTP(file *os.File, URL *url.URL) error {
+	if URL.User == nil {
+		return errors.New("No User Account or password provided")
+	}
+
+	keyFile, err := os.Open(env.CI_SSH_PUBLIC_KEY)
+	if err != nil {
+		return err
+	}
+
+	privateKey, err := io.ReadAll(keyFile)
+	if err != nil {
+		return err
+	}
+
+	var Signer ssh.Signer
+
+	if sshPW := os.Getenv(env.ENV_SSH_PRIVATE_PW); sshPW != "" {
+		Signer, err = ssh.ParsePrivateKeyWithPassphrase(privateKey, []byte(sshPW))
+	} else {
+		Signer, err = ssh.ParsePrivateKey(privateKey)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	AuthMethods := []ssh.AuthMethod{ssh.PublicKeys(Signer)}
+
+	var hostPublicKey ssh.PublicKey
+
+	if key := os.Getenv(env.ENV_SSH_HOST_PUBLIC); key != "" {
+		hostPublicKey, err = ssh.ParsePublicKey([]byte(key))
+		if err != nil {
+			return err
+		}
+	}
+
+	sshClient, err := ssh.Dial("tcp", URL.Host, &ssh.ClientConfig{
+		User:            URL.User.Username(),
+		Auth:            AuthMethods,
+		HostKeyCallback: ssh.FixedHostKey(hostPublicKey),
+	})
+	if err != nil {
+		return err
+	}
+	defer sshClient.Close()
+
+	{
```

Comment:
Are the curly brackets needed here?

Reply:
I just tested this and the answer is very likely no :D